ext/FFTWChainRulesCoreExt.jl

```diff
@@ -0,0 +1,72 @@
+module FFTWChainRulesCoreExt
+
+using FFTW
+using FFTW: r2r
+using ChainRulesCore
+
+# DCT
+
+function ChainRulesCore.frule(Δ, ::typeof(dct), x::AbstractArray, region...)
+    Δx = Δ[2]
+    y = dct(x, region...)
+    Δy = dct(Δx, region...)
+    return y, Δy
+end
+
+function ChainRulesCore.rrule(::typeof(dct), x::AbstractArray, region...)
```

Comment:
Same here, would have been nice to be able to just use `region` or `region=1:ndims(x)` - also to avoid the branch in the pullback below.

Reply:
You are right, the call sequence is a bit weird. I was attempting to get away with defining a single `rrule` for both `dct(x)` and `dct(x, region)`.

In FFTW.jl, if `dct(x)` were set to `dct(x, 1:ndims(x))`, then we could get away with defining an rrule on `dct(x, region)`. But `dct(x)` actually calls `plan_dct(x) * x`.

I tried your suggestion of setting `region = 1:ndims(x)` in the `rrule`, but then either `dct(x, region)` or `dct(x)` fail with below error depending on whether I return a cotangent for `region`.
```julia
test_rrule: dct on Matrix{Float64}: Test Failed at /Users/vp/.julia/packages/ChainRulesTestUtils/C9L2i/src/testers.jl:226                                                                                       
  Expression: length(ad_cotangents) == length(primals)                                                                                                                                                          
  Problem: The pullback should return 1 cotangent for the primal and each primal input.                                                                                                                         
   Evaluated: 3 == 2                                                                                                                                                                                            
```

Reply:
It's easier to write different rules for `dct(x)`, `dct(x, region)`. That is done in https://github.com/JuliaMath/FFTW.jl/pull/273/commits/6b8d178afd368f1b5e5334b76e8b9b617fc3e6d6.